odl/space/npy_tensors.py

```diff
@@ -0,0 +1,2336 @@
+        - A distance function or metric on a space :math:`\mathcal{X}`
+          is a mapping
+          :math:`d:\mathcal{X} \\times \mathcal{X} \\to \mathbb{R}`
+          satisfying the following conditions for all space elements
+          :math:`x, y, z`:
+
+          * :math:`d(x, y) \geq 0`,
+          * :math:`d(x, y) = 0 \Leftrightarrow x = y`,
+          * :math:`d(x, y) = d(y, x)`,
+          * :math:`d(x, y) \\leq d(x, z) + d(z, y)`.
+
+        - A norm on a space :math:`\mathcal{X}` is a mapping
+          :math:`\| \cdot \|:\mathcal{X} \\to \mathbb{R}`
+          satisfying the following conditions for all
+          space elements :math:`x, y`: and scalars :math:`s`:
+
+          * :math:`\| x\| \geq 0`,
+          * :math:`\| x\| = 0 \Leftrightarrow x = 0`,
+          * :math:`\| sx\| = |s| \cdot \| x \|`,
+          * :math:`\| x+y\| \\leq \| x\| +
+            \| y\|`.
+
+        - An inner product on a space :math:`\mathcal{X}` over a field
+          :math:`\mathbb{F} = \mathbb{R}` or :math:`\mathbb{C}` is a
+          mapping
+          :math:`\\langle\cdot, \cdot\\rangle: \mathcal{X} \\times
+          \mathcal{X} \\to \mathbb{F}`
+          satisfying the following conditions for all
+          space elements :math:`x, y, z`: and scalars :math:`s`:
+
+          * :math:`\\langle x, y\\rangle =
+            \overline{\\langle y, x\\rangle}`,
+          * :math:`\\langle sx + y, z\\rangle = s \\langle x, z\\rangle +
+            \\langle y, z\\rangle`,
+          * :math:`\\langle x, x\\rangle = 0 \Leftrightarrow x = 0`.
+
+        Examples
+        --------
+        >>> space = NumpyTensorSpace((2, 3), dtype='float64')
+        >>> space
+        rn((2, 3))
+
+        >>> space = NumpyTensorSpace((2, 3), dtype='complex64')
+        >>> space
+        cn((2, 3), 'complex64')
+
+        >>> space = NumpyTensorSpace((2, 3), dtype='int64')
+        >>> space
+        tensor_space((2, 3), 'int')
+        """
+        TensorSpace.__init__(self, shape, dtype, order)
+
+        dist = kwargs.pop('dist', None)
+        norm = kwargs.pop('norm', None)
+        inner = kwargs.pop('inner', None)
+        weighting = kwargs.pop('weighting', None)
+        exponent = kwargs.pop('exponent', 2.0)
+        dist_using_inner = bool(kwargs.pop('dist_using_inner', False))
+
+        if (not self.is_numeric and
+                any(x is not None for x in (dist, norm, inner, weighting))):
+            raise ValueError('cannot use any of `weighting`, `dist`, `norm` '
+                             'or `inner` for non-numeric `dtype` {}'
+                             ''.format(dtype))
+        if exponent != 2.0 and any(x is not None for x in (dist, norm, inner)):
+            raise ValueError('cannot use any of `dist`, `norm` or `inner` '
+                             'for exponent != 2')
+        # Check validity of option combination (3 or 4 out of 4 must be None)
+        if sum(x is None for x in (dist, norm, inner, weighting)) < 3:
```

Comment:
I'm not really fond of code that relies on `None` evaluating to `False`. That's like statements
```python
impl = str(impl).lower() or 'none'
```
instead of 
```python
impl = 'none' if impl is None else str(impl).lower()
```
I find the latter more readable and more explicit.